gremlin-go/driver/connection.go

```diff
@@ -52,7 +60,7 @@ func (connection *connection) errorCallback() {
 
 func (connection *connection) close() error {
 	if connection.state != established {
-		return errors.New("cannot close connection that has already been closed or has not been connected")
+		return newError(err0101ConnectionCloseError)
 	}
```

Comment:
We need to introduce a 'closing' state here. For the same connection, this method could be called simultaneously from two threads (e.g. one from go routine inside pool.write and another from pool.close). Only one should be allowed to 'close'.

Also please add method doc which mentions whether a function is idempotent or not and also whether it is thread safe or not.